src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -306,29 +273,51 @@ public static ImmutableArray<DocumentSymbolUIItem> GetDocumentSymbolUIItems(Immu
+        /// Updates the IsExpanded property for the Document Symbol ViewModel based on the given Expansion Option. The parameter
         /// <param name="currentDocumentSymbolItems"/> is used to reference the current node expansion in the view.
         /// </summary>
-        public static void SetIsExpanded(
-            ImmutableArray<DocumentSymbolUIItem> documentSymbolItems,
-            IEnumerable<DocumentSymbolUIItem> currentDocumentSymbolItems,
+        public static void SetIsExpandedOnNewItems(
+            ImmutableArray<DocumentSymbolDataViewModel> newDocumentSymbolItems,
+            IEnumerable<DocumentSymbolDataViewModel> currentDocumentSymbolItems)
+        {
+            using var _ = PooledHashSet<DocumentSymbolDataViewModel>.GetInstance(out var hashset);
+            hashset.AddRange(newDocumentSymbolItems);
+
+            foreach (var item in currentDocumentSymbolItems)
+            {
+                if (!hashset.TryGetValue(item, out var newItem))
+                {
+                    continue;
+                }
+
+                newItem.IsExpanded = item.IsExpanded;
+                SetIsExpandedOnNewItems(newItem.Children, item.Children);
+            }
+        }
+
+        public static void SetExpansionOption(
+            IEnumerable<DocumentSymbolDataViewModel> currentDocumentSymbolItems,
             ExpansionOption expansionOption)
         {
-            for (var i = 0; i < documentSymbolItems.Length; i++)
+            foreach (var item in currentDocumentSymbolItems)
             {
-                if (expansionOption is ExpansionOption.CurrentExpansion)
-                    documentSymbolItems[i].IsExpanded = currentDocumentSymbolItems.ElementAt(i).IsExpanded;
-                else
-                    documentSymbolItems[i].IsExpanded = expansionOption is ExpansionOption.Expand;
+                if (expansionOption is ExpansionOption.Collapse)
+                {
+                    item.IsExpanded = false;
+                }
+                else if (expansionOption is ExpansionOption.Expand)
+                {
+                    item.IsExpanded = true;
+                }
 
-                SetIsExpanded(documentSymbolItems[i].Children, currentDocumentSymbolItems.ElementAt(i).Children, expansionOption);
+                SetExpansionOption(item.Children, expansionOption);
             }
         }
 
         /// <summary>
-        /// Expands all the ancestors of a DocumentSymbolUIItem.
+        /// Expands all the ancestors of a <see cref="DocumentSymbolDataViewModel"/>.
         /// </summary>
-        public static void ExpandAncestors(ImmutableArray<DocumentSymbolUIItem> documentSymbolItems, SnapshotSpan documentSymbolRangeSpan)
+        public static void ExpandAncestors(IEnumerable<DocumentSymbolDataViewModel> documentSymbolItems, SnapshotSpan documentSymbolRangeSpan)
```

Comment:
passing around IEnumerables always worries me as they may represent lazy computation, which means it's hard to ascertain if you get the same values back.  for code that is doing mutation, it's preferable to use concrete collection types or interfaces.